Knife Profile

kagarmoe opened the pull request:
### Description

Adds @btm's content for Knife Profiles
### Definition of Done

### Issues Resolved

Fixes #1603

### Check List

- [x] Spell Check
- [x] Local build
- [x] Examine the local build
- [x] All tests pass

Commit:
Knife Profile

btm commented on chef_master/source/knife_setup.rst:
```diff
@@ -58,6 +58,77 @@ Generating a config.rb File
+1. Write the profile name to the `~/.chef/context` file.
+1. Otherwise, knife will use the 'default' profile.
+
+knife config support since Chef 14.4
```
This line can be removed, as the information is in the line below. This probably should be a minor header of some kind for 'knife config' though.

btm commented on chef_master/source/knife_setup.rst:
```diff
@@ -58,6 +58,77 @@ Generating a config.rb File
+Knife Profiles
+=====================================================
+
+Profile support since Chef 13.7
```
I don't know if there's a tag or pattern for this?

mjingle replied:
`**Profile Support since Chef 13.7**`

mbaitelman commented:
Not sure how related this is but I was testing this out after @btm 's session and this doesnt play nicely with the `--config CONFIG_FILE` option for https://docs.chef.io/policyfile.html#chef-commands. 
Will the `~/.chef/credentials` file work with `chef` commands?

mbaitelman commented on chef_master/source/knife_setup.rst:
```diff
@@ -58,6 +58,77 @@ Generating a config.rb File
+# Example .chef/credentials file
+[default]
+node_name = "barney"
+client_key = "barney_rubble.pem"
```
Could you add an example inline key?
```suggestion
client_key = """
-----BEGIN RSA PRIVATE KEY-----
MIICXAIBAAKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0FPqri0cb2JZfXJ/DgYSF6vUp
wmJG8wVQZKjeGcjDOL5UlsuusFncCzWBQ7RKNUSesmQRMSGkVb1/3j+skZ6UtW+5u09lHNsj6tQ5
1s1SPrCBkedbNf0Tp0GbMJDyR4e9T04ZZwIDAQABAoGAFijko56+qGyN8M0RVyaRAXz++xTqHBLh
3tx4VgMtrQ+WEgCjhoTwo23KMBAuJGSYnRmoBZM3lMfTKevIkAidPExvYCdm5dYq3XToLkkLv5L2
pIIVOFMDG+KESnAFV7l2c+cnzRMW0+b6f8mR1CJzZuxVLL6Q02fvLi55/mbSYxECQQDeAw6fiIQX
GukBI4eMZZt4nscy2o12KyYner3VpoeE+Np2q+Z3pvAMd/aNzQ/W9WaI+NRfcxUJrmfPwIGm63il
AkEAxCL5HQb2bQr4ByorcMWm/hEP2MZzROV73yF41hPsRC9m66KrheO9HPTJuo3/9s5p+sqGxOlF
L0NDt4SkosjgGwJAFklyR1uZ/wPJjj611cdBcztlPdqoxssQGnh85BzCj/u3WqBpE2vjvyyvyI5k
X6zk7S0ljKtt2jny2+00VsBerQJBAJGC1Mg5Oydo5NwD6BiROrPxGo2bpTbu/fhrT8ebHkTz2epl
U9VQQSQzY1oZMVX8i1m5WUTLPz2yLJIBQVdXqhMCQBGoiuSoSjafUhV7i1cEGpb88h5NBYZzWXGZ
37sJ5QsW+sJyoNde3xH8vdXhzU7eT82D6X/scw9RZz+/6rCJ4p0=
-----END RSA PRIVATE KEY-----
"""
```

btm replied:
added, thanks!

btm committed:
Add additional knife profile examples

- Fix 'since' lines
- add example for an inline key
- add example for knife config context

Signed-off-by: Bryan McLellan <btm@loftninjas.org>

btm commented:
@mbainter I'm not sure if `chef` supports a credentials file. it does appear to use `Chef::WorkstationConfigLoader` so it could. Trying to combine `--config foo.rb` with a credentials file feels like you're going to end up confused when they merge, even if it did work. What are you trying to configure?

mbaitelman commented:
@btm you are correct. chef commands do not support the credentials file. 
I had assumed that it would work. 
The use case here was to have one file that would have all settings for connecting to my chef server.
This would be helpful in my CI/CD tooling that runs in containers and I have to write the files each time. 
Also using the credentials as an ENV var vs a local file is cleaner and simpler to maintain. 
I will create an issue in chef/chef to see about adding support for credentials in chef commands.

Commit:
Sphinx formatting and some copyediting

Signed-off-by: Mary Jinglewski <mjinglewski@chef.io>

mjingle commented on chef_master/source/knife_setup.rst:
```diff
@@ -56,7 +56,103 @@ Example config.rb:
+    3. Write the profile name to the ``~/.chef/context`` file.
+    4. Otherwise, knife will use the 'default' profile.
+
+Knife Config
```
@btm Would it be safe to say that this Knife Config deserves its own header/section?

btm replied:
@mjingle yeah, it's not complete dependent on profiles, although most of it's functionality is profile related.

mjingle reviewed (approved):
![tenor-108656121](https://user-images.githubusercontent.com/5807053/58907579-82638c80-86dc-11e9-83ea-7ad299dbe033.gif)
